contrib/vendor-bitcoin-core.sh

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env bash
+set -e
+
+CORE_VENDOR_VERSION=""          # bitcoin repo tagged version to vendor.
+
+DEFAULT_DEPEND_DIR="depend"
+DEFAULT_CORE_REPO=https://github.com/bitcoin/bitcoin.git
+
+: "${CORE_VENDOR_DEPEND_DIR:=$DEFAULT_DEPEND_DIR}"
+: "${CORE_VENDOR_REPO:=$DEFAULT_CORE_REPO}"
+
+# CP_NOT_CLONE lets us just copy a directory rather than git cloning.
+# This is usually a bad idea, since it will bring in build artifacts or any other
+# junk from the source directory, but may be useful during development or CI.
+: "${CORE_VENDOR_CP_NOT_CLONE:=no}"
+
+function usage() {
+    echo ""
+    echo "Usage: script <bitcoin-core-version-tag>"
+    echo ""
+    echo -ne '\t'
+    echo "$0 v0.21.2"
+    echo ""
+}
+
+if $# < 1; then
+   usage
+fi
+
+echo "Using depend directory $CORE_VENDOR_DEPEND_DIR. Set CORE_VENDOR_DEPEND_DIR to override."
+echo "Using bitcoin repository $CORE_VENDOR_REPO. Set CORE_VENDOR_REPO to override."
+
+# Parse command-line options
+FORCE=no
+while (( "$#" )); do
+    case "$1" in
+    -f)
+        FORCE=yes
+        ;;
+    *)
+        if [ -z "$CORE_VENDOR_VERSION" ]; then
+            CORE_VENDOR_VERSION="$1"
+        else
+            echo "WARNING: ignoring unknown command-line argument $1"
+        fi
+        ;;
+    esac
+    shift
+done
+
+echo "Vendoring Bitcoin Core version: $CORE_VENDOR_VERSION"
+
+# Check if we will do anything destructive.
+
+if [ "$FORCE" == "no" ]; then
+    if ! git diff --quiet -- "*.rs"; then
+        echo "ERROR: There appear to be modified source files. Check these in or pass -f (some source files will be modified to have symbols renamed)."
+        exit 2
+    fi
+    if ! git diff --quiet -- "$CORE_VENDOR_DEPEND_DIR"; then
+        echo "ERROR: The depend directory appears to be modified. Check it in or pass -f (this directory will be deleted)."
+        exit 2
+    fi
+fi
+
+DIR=./bitcoin
+
+pushd "$CORE_VENDOR_DEPEND_DIR" > /dev/null
+rm -rf "$DIR" || true
+
+# Clone the repo. As a special case, if the repo is a local path and we have
+# not specified a revision, just copy the directory rather than using 'git clone'.
+# This lets us use non-git repos or dirty source trees as secp sources.
+if [ "$CORE_VENDOR_CP_NOT_CLONE" == "yes" ]; then
+    cp -r "$CORE_VENDOR_REPO" .
```

Comment:
In b83e1debf3f2c24985ba37db0d29f326c21b179e:

In the rust-secp version of this it's a `cp "$VENDOR_REPO" "$DIR"`. Why did you change `$DIR` to `.`? It causes the following `chmod` to fail.

Reply:
At the time I couldn't work out a sane way to call the script, I was using 
```bash
 CORE_VENDOR_CP_NOT_CLONE=yes CORE_VENDOR_REPO=/home/tobin/build/github.com/tcharding/bitcoin contrib/vendor-bitcoin-core.sh v0.21-final
```

But that results in `bitcoin/bitcoin`?

Reply:
Are you using ` CORE_VENDOR_REPO=/path/to/bitcoin/*`

Reply:
No, just `CORE_VENDOR_REPO=/path/to/bitcoin`. As in

```
CORE_VENDOR_CP_NOT_CLONE=yes \
CORE_VENDOR_GIT_ROOT=.. \
CORE_VENDOR_REPO=/nix/store/77qhrhjq230jh6zswzg2nfxirpw4zdd3-source \
CORE_VENDOR_DEPEND_DIR=./depend2/ \
./contrib/vendor-bitcoin-core.sh -f
```

Reply:
If you are not specifying `CORE_VENDOR_DEPEND_DIR` then it will default to `depend/`. And you'll get a directory `depend/bitcoin` as expected. I'm not sure how you would get `bitcoin/bitcoin`.

Reply:
This is the verdor dir not the depend dir.

Reply:
This is the offending line
```
CORE_VENDOR_REPO=/nix/store/77qhrhjq230jh6zswzg2nfxirpw4zdd3-source
```
that directory is not called `bitcoin` like when I call it.

Reply:
I'll change it to how you use it, I can see how this is good for your usecase, when I do  `~/build/github.com` it doesn't seem good. But its a trifling matter.

Reply:
When you use `.` as the destination then it matters what the source directory is called because you are copying it into `.`. When you use a specific name, then you are copying the source directory to the new name (not *into* it) so it has exactly the same effect except that the source name doesn't matter. So I don't understand why this is worse for you.

Reply:
ooooo, TIL, if the destination directory exists already then the behaviour of `cp` is different to if it does not exist - I have never realized this, wow, mind blown. If the directory _does not exist_ it is as you say, it creates the directory correctly, but if it exists then it gets copied _into_ the existing one.

Reply:
I still don't know how this was happening though because we explicitly `rm -rf $DIR` right above. Anyways, onwards and upwards!

Reply:
> I have never realized this, wow, mind blown

FWIW I always use `rsync` and never use `cp -r`,  its still surprising that I did not know the nuances of `cp`.

Reply:
I wouldn't be *opposed* to using rsync here, FWIW, though it would be an extra dependency I'd have to pull into my nix derivation and it's probably overkill here.

Agreed that `cp` is nuts :)

Reply:
Oh I didn't mean to use it here, I just meant on my machine I never use `cp` for directories.

Reply:
OMG, one last (hopefully) absolute face palm for the year
```
CORE_VENDOR_REPO=/home/tobin/build/github.com/tcharding/bitcoin
``` 

`github.com/bitcoin` is the org name, it contains the repostory `bitcoin` - it was not `cp`'s fault at all, it was all mine.